// nativescript/lib/commands/extensibility/uninstall-extension.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.UninstallExtensionCommand = void 0;
const yok_1 = require("../../common/yok");
class UninstallExtensionCommand {
    constructor($extensibilityService, $stringParameterBuilder, $logger) {
        this.$extensibilityService = $extensibilityService;
        this.$stringParameterBuilder = $stringParameterBuilder;
        this.$logger = $logger;
        this.allowedParameters = [
            this.$stringParameterBuilder.createMandatoryParameter("You have to provide a valid name for extension that you want to uninstall."),
        ];
    }
    execute(args) {
        return __awaiter(this, void 0, void 0, function* () {
            const extensionName = args[0];
            yield this.$extensibilityService.uninstallExtension(extensionName);
            this.$logger.info(`Successfully uninstalled extension ${extensionName}`);
        });
    }
}
exports.UninstallExtensionCommand = UninstallExtensionCommand;
yok_1.injector.registerCommand("extension|uninstall", UninstallExtensionCommand);
